File: back/src/main/java/csh/back/domain/trip/timeline/dto/response/TimeLineWithVoteIdResponse.java
package csh.back.domain.trip.timeline.dto.response;

import csh.back.domain.trip.place.entity.TripPlace;
import csh.back.domain.trip.timeline.entity.TimeLine;

import java.time.LocalDateTime;

public record TimeLineWithVoteIdResponse(
        Long timelineId,
        Integer dayNumber,
        LocalDateTime startTime,
        LocalDateTime endTime,
        String confirmedPlaceName,
        String category,
        Long voteId
) {
    //서비스 로직에서 조회 결과를 응답으로 바꿀 때 편하게 하기 위한 정적 메서드
    public static TimeLineWithVoteIdResponse of(TimeLine timeLine, Long voteId) {
        String confirmedPlaceName = timeLine.getConfirmedPlace() != null ? timeLine.getConfirmedPlace().getName() : null;
        String category = timeLine.getConfirmedPlace() != null ? timeLine.getConfirmedPlace().getTheme() : null;
        return new TimeLineWithVoteIdResponse(
                timeLine.getId(),
                timeLine.getDayNumber(),
                timeLine.getStartTime(),
                timeLine.getEndTime(),
                confirmedPlaceName,
                category,
                voteId
        );
    }
}

File: back/src/main/kotlin/csh/back/domain/trip/timeline/dto/response/TimelineWithVoteIdResponse.kt
package csh.back.domain.trip.timeline.dto.response

import csh.back.domain.trip.timeline.entity.Timeline
import java.time.LocalDateTime

data class TimelineWithVoteIdResponse(
    val timelineId: Long,
    val dayNumber: Long,
    val startTime: LocalDateTime,
    val endTime: LocalDateTime,
    val confirmedPlaceName: String?,
    val category: String?,
    val voteId: Long?,
    val isFreeTime: Boolean,
) {
    companion object {
        @JvmStatic
        fun of(timeline: Timeline, voteId: Long?): TimelineWithVoteIdResponse {
            val tripWishPlace = timeline.tripWishPlace

            return TimelineWithVoteIdResponse(
                timelineId = requireNotNull(timeline.id),
                dayNumber = timeline.dayNumber,
                startTime = timeline.startTime,
                endTime = timeline.endTime,
                confirmedPlaceName = if (timeline.isFreeTime) FREE_TIME_NAME else tripWishPlace?.name,
                category = tripWishPlace?.category,
                voteId = if (timeline.isFreeTime) null else voteId,
                isFreeTime = timeline.isFreeTime,
            )
        }

        private const val FREE_TIME_NAME = "자유시간"
    }
}
